explorer/interpreter/type_checker.cpp

```diff
@@ -4521,6 +4521,25 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
              << "Error declaring `" << fun->name() << "`"
              << ": class functions cannot be virtual.";
     }
+    if (fun->override() == VirtualOverride::Abstract) {
+      return ProgramError(fun->source_loc())
+             << "Error declaring `" << fun->name() << "`"
+             << ": `abstract` methods are not yet supported.";
+    }
+    if (fun->override() == VirtualOverride::Virtual &&
+        class_vtable.find(fun->name()) != class_vtable.end()) {
+      return ProgramError(fun->source_loc())
+             << "Error declaring `" << fun->name() << "`"
+             << ": method is declared virtual in base class, use `impl` "
+                "instead to override it.";
+    }
+    if (fun->override() == VirtualOverride::Impl &&
+        class_vtable.find(fun->name()) == class_vtable.end()) {
+      return ProgramError(fun->source_loc())
+             << "Error declaring `" << fun->name() << "`"
+             << ": cannot override a method that is not declared `abstract` or "
+                "`virtual` in base class.";
+    }
```

Comment:
```suggestion
    bool has_vtable_entry = class_vtable.find(fun->name()) != class_vtable.end();
    switch (fun->override()) {
      case VirtualOverride::Abstract:
      case VirtualOverride::None:
      case VirtualOverride::Virtual:
      if (has_vtable_entry) {
      return ProgramError(fun->source_loc())
             << "Error declaring `" << fun->name() << "`"
             << ": method is declared virtual in base class, use `impl` "
                "to override it.";
      }
      if (fun->override() == VirtualOverride::None) {
        // Not added to the vtable.
        return;
      }
      break;
      case VirtualOverride::Impl:
      if (!has_vtable_entry) {
      return ProgramError(fun->source_loc())
             << "Error declaring `" << fun->name() << "`"
             << ": cannot override a method that is not declared `abstract` or "
                "`virtual` in base class.";
      }
      break;
    }
```

Suggesting the slight restructuring -- I think you're going to end up with all the virtual override cases (including abstract and none) wanting to do the vtable find, so do it once.

I've added Abstract and None to the above. The Abstract code is mainly an example, given the above `if`; I have a comment above regarding None. I think this is part of the value of `switch` though, in that it seems simpler when adding more cases.

No default because then clang will check the enum is fully handled (another `switch` advantage).

Reply:
Thanks for taking the effort to write it all down :). Applied. As discussed, I don't know if this will work with the check matrix from the design doc, but this is good for now, and at least for now, better organized.